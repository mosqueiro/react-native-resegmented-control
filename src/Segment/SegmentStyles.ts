import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    zIndex: 2,
  },
  disabled: {
    opacity: 0.5,
  },
  segment: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  segmentText: {
    fontSize: 13,
    paddingLeft: 2,
    paddingRight: 2,
    width: '100%',
    textAlign: 'center',
  },
  segmentActiveText: {
    fontWeight: 'bold',
  },
});

export default styles;
