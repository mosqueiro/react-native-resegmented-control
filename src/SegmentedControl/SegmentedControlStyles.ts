import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#eeeeef',
    flexDirection: 'row',
    alignItems: 'center',
    borderRadius: 8,
    height: 28,
    position: 'relative',
  },
  slider: {
    position: 'absolute',
    top: 0,
    left: 0,
    zIndex: 1,
  },
  sliderDefault: {
    height: '86%',
    backgroundColor: 'white',
    borderRadius: 7,
    margin: 2,
    shadowOffset: { width: 0.95, height: 0.95 },
    shadowColor: '#a2a2a2',
    shadowOpacity: 0.5,
    shadowRadius: 2,
  },
  dividerContainer: {
    paddingTop: 7,
    paddingBottom: 7,
    zIndex: 0,
  },
  divider: {
    height: '100%',
    width: 1,
    borderWidth: 0,
    backgroundColor: 'rgba(120, 120, 120, 0.2)',
  },
});

export default styles;
