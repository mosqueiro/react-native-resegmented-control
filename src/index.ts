export { SegmentedControl, SegmentedControlProps } from './SegmentedControl';
export { Segment, SegmentProps, SegmentContentProps } from './Segment';
