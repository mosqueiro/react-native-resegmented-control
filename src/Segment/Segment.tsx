import React, { useContext, FC } from 'react';
import { StyleProp, Text, View, ViewStyle } from 'react-native';
import { TouchableOpacity } from 'react-native-gesture-handler';
import { SegmentedContext } from '../SegmentedControl/SegmentedControl';
import styles from './SegmentStyles';

export interface SegmentProps {
  activeTintColor?: string;
  content: React.ReactNode;
  inactiveTintColor?: string;
  name: string;
  style?: StyleProp<ViewStyle>;
}

export const Segment: FC<SegmentProps> = ({
  activeTintColor,
  content,
  inactiveTintColor,
  name,
  style,
}: SegmentProps) => {
  const context = useContext(SegmentedContext);

  if (!context) {
    throw new Error('Segment must be used within a SegmentedControl.');
  }

  const { selectedName, onChange } = context;

  const active = selectedName === name;

  const handlePress = (): void => {
    if (typeof onChange === 'function') {
      onChange(name);
    }
  };

  const renderContent = (): React.ReactNode => {
    if (
      typeof content === 'string' ||
      typeof content === 'number' ||
      typeof content === 'boolean'
    ) {
      return (
        <Text
          numberOfLines={1}
          style={[
            styles.segmentText,
            active ? { color: activeTintColor } : { color: inactiveTintColor },
            active && styles.segmentActiveText,
          ]}
        >
          {content}
        </Text>
      );
    }

    if (typeof content === 'function') {
      return content({ activeTintColor, inactiveTintColor, active });
    }

    return content;
  };

  return (
    <View style={[styles.container, style as ViewStyle]}>
      <TouchableOpacity onPress={handlePress} testID={`Segment_Button`}>
        <View style={styles.segment}>{renderContent()}</View>
      </TouchableOpacity>
    </View>
  );
};
