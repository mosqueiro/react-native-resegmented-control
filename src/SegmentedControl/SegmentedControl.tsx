import React, { createContext, useEffect, useState } from 'react';
import { LayoutChangeEvent, View, ViewStyle } from 'react-native';
import { PanGestureHandler, PanGestureHandlerGestureEvent } from 'react-native-gesture-handler';
import Animated, { Easing } from 'react-native-reanimated';
import { timing } from 'react-native-redash';

import { Segment, SegmentProps } from '../Segment';
import { clamp } from '../utils';
import styles from './SegmentedControlStyles';

export interface SegmentedControlProps {
  activeTintColor?: string;
  inactiveTintColor?: string;
  initialSelectedName?: string;
  children:
    | React.ReactElement<SegmentProps>
    | React.ReactElement<SegmentProps>[];
  onChangeValue?: (name: string) => void;
  sliderStyle?: ViewStyle;
  style?: ViewStyle;
}

export const SegmentedContext = createContext<{
  selectedName: string | null | undefined;
  onChange: ((name: string) => void) | undefined;
} | null>(null);

export const SegmentedControl = ({
  activeTintColor = '#000000',
  children,
  inactiveTintColor = '#000000',
  initialSelectedName,
  onChangeValue,
  sliderStyle,
  style,
}: SegmentedControlProps): JSX.Element => {
  const [_initialized, _setInitialized] = useState(false);
  const [_width, _setWidth] = useState(0);
  const [_initialSelectedName] = useState(initialSelectedName);
  const [_activeName, _setActiveName] = useState(_initialSelectedName);
  const [_sliderPosition, _setSliderPosition] = useState(
    new Animated.Value<number>(0),
  );
  const [_sliderWidth, _setSliderWidth] = useState(0);
  const [_map, _setMap] = useState<{ [key: string]: number } | undefined>(
    undefined,
  );

  const values = Array.isArray(children) ? children : [children];

  // Map segment names to index
  useEffect(() => {
    const tempMap = {};

    values.forEach((child, index) => {
      if (child.type !== Segment) {
        throw new Error('SegmentedControl only accepts Segment as children.');
      }

      if (!child.props.name) {
        throw new Error('Segment requires `name` to be defined.');
      }

      tempMap[child.props.name] = index;
    });

    _setMap(tempMap);
  }, []);

  // Set slider width
  useEffect(() => {
    _setSliderWidth(_width * (1 / values.length - 0.015));
  }, [values, _width]);

  // Set initial slider position
  useEffect(() => {
    if (
      typeof _initialSelectedName !== 'undefined' &&
      typeof _map !== 'undefined' &&
      _width > 0 &&
      !_initialized
    ) {
      const index = _map[_initialSelectedName];
      const position = _width * (index / values.length);
      _setSliderPosition(new Animated.Value(position));
      _setInitialized(true);
    }
  }, [values, _width, _map, _initialSelectedName]);

  // This hook is used to animate the slider position
  Animated.useCode(() => {
    const index = _activeName && _map ? _map[_activeName] : 0;
    const sliderPosition = _width * (index / values.length);

    return Animated.set(
      _sliderPosition,
      timing({
        from: _sliderPosition,
        to: sliderPosition,
        easing: Easing.linear,
        duration: 200,
      }),
    );
  }, [_activeName]);

  const handleLayout = (event: LayoutChangeEvent): void =>
    _setWidth(event.nativeEvent.layout.width);

  const handleChangeValue = (name: string): void => {
    if (typeof _activeName === 'undefined' && typeof _map !== 'undefined') {
      const index = _map[name];
      _setSliderPosition(new Animated.Value(_width * (index / values.length)));
    }
    _setActiveName(name);

    if (typeof onChangeValue === 'function') {
      onChangeValue(name);
    }
  };

  const handleGestureEvent = (event: PanGestureHandlerGestureEvent): void => {
    const { x } = event.nativeEvent;

    const calculatedIndex = Math.floor((x / _width) * values.length);
    const index = clamp(calculatedIndex, 0, values.length - 1);
    const { name } = values[index].props;

    handleChangeValue(name);
  };

  return (
    <SegmentedContext.Provider
      value={{ selectedName: _activeName, onChange: handleChangeValue }}
    >
      <PanGestureHandler onGestureEvent={handleGestureEvent}>
        <View onLayout={handleLayout} style={[styles.container, style]}>
          {typeof _activeName !== 'undefined' && (
            <Animated.View
              testID="SegmentedControl_Slider"
              style={[
                styles.sliderDefault,
                {
                  width: _sliderWidth,
                  transform: [
                    {
                      translateX: _sliderPosition,
                    },
                  ],
                },
                sliderStyle,
                styles.slider,
              ]}
            />
          )}

          {values.map((child, index) => (
            <React.Fragment key={child.props.name}>
              {index > 0 && (
                <View style={styles.dividerContainer}>
                  <View style={styles.divider} />
                </View>
              )}
              {{
                ...child,
                props: {
                  ...child.props,
                  inactiveTintColor,
                  activeTintColor,
                },
              }}
            </React.Fragment>
          ))}
        </View>
      </PanGestureHandler>
    </SegmentedContext.Provider>
  );
};

export default SegmentedControl;
