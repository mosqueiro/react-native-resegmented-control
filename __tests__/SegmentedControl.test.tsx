import React from 'react';
import { View } from 'react-native';
import { fireEvent, render } from '@testing-library/react-native';
import { SegmentedControl } from '../src/SegmentedControl';
import { Segment } from '../src/Segment';

jest.mock('react-native-reanimated', () =>
  require('react-native-reanimated/mock'),
);

jest.mock('react-native-gesture-handler', () => {
  const actual = jest.requireActual('react-native-gesture-handler');
  const { TouchableOpacity } = jest.requireActual('react-native');
  return {
    ...actual,
    TouchableOpacity,
  };
});

describe('SegmentedControl', () => {
  it('should render', () => {
    const { asJSON } = render(
      <SegmentedControl>
        <Segment name="Test" content="Test" />
      </SegmentedControl>,
    );

    expect(asJSON()).toMatchSnapshot();
  });

  it('should render initially without slider, press on a segment and slider should appear', async () => {
    const { getByTestId, getByText } = render(
      <SegmentedControl>
        <Segment name="first" content="First" />
        <Segment name="second" content="Second" />
      </SegmentedControl>,
    );

    expect(() => getByTestId('SegmentedControl_Slider')).toThrow();

    const secondSegment = getByText('Second');
    fireEvent.press(secondSegment);

    expect(getByTestId('SegmentedControl_Slider')).toBeDefined();
  });

  it('should render initially with slider on `Second`', async () => {
    const { getByTestId } = render(
      <SegmentedControl initialSelectedName="second">
        <Segment name="first" content="First" />
        <Segment name="second" content="Second" />
      </SegmentedControl>,
    );

    expect(getByTestId('SegmentedControl_Slider')).toBeDefined();
  });

  it('should call onChangeValue when pressed on `Test`', async () => {
    const changeValueSpy = jest.fn();
    const { getByTestId } = render(
      <SegmentedControl onChangeValue={changeValueSpy}>
        <Segment name="Test" content="Test" />
      </SegmentedControl>,
    );

    const button = getByTestId('Segment_Button');
    fireEvent.press(button);

    expect(changeValueSpy).toBeCalledWith('Test');
  });

  it('should throw if a child component is not a Segment', () => {
    const errorSpy = jest.spyOn(console, 'error');
    errorSpy.mockImplementation();

    expect(() => {
      render(
        <SegmentedControl>
          <View />
        </SegmentedControl>,
      );
    }).toThrow('SegmentedControl only accepts Segment as children.');

    errorSpy.mockRestore();
  });

  it('should throw if a Segment has no name', () => {
    const errorSpy = jest.spyOn(console, 'error');
    errorSpy.mockImplementation();

    expect(() => {
      render(
        <SegmentedControl>
          <Segment name="" content="No Name" />
        </SegmentedControl>,
      );
    }).toThrow('Segment requires `name` to be defined.');

    errorSpy.mockRestore();
  });
});
