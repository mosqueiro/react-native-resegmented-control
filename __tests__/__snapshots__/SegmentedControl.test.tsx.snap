// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentedControl should render 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={false}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#eeeeef",
          "borderRadius": 8,
          "flexDirection": "row",
          "height": 28,
          "position": "relative",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "zIndex": 2,
          },
          undefined,
        ]
      }
    >
      <TouchableOpacity
        activeOpacity={0.2}
        testID="Segment_Button"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "fontSize": 13,
                  "paddingLeft": 2,
                  "paddingRight": 2,
                  "textAlign": "center",
                  "width": "100%",
                },
                Object {
                  "color": "#000000",
                },
                false,
              ]
            }
          >
            Test
          </Text>
        </View>
      </TouchableOpacity>
    </View>
  </View>
</View>
`;
